src/tools/rbd_mirror/image_map/ImageAction.h

```diff
@@ -0,0 +1,138 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+
+#include <map>
+#include <list>
+#include <string>
+
+#include "Types.h"
+#include "common/Mutex.h"
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+class ImageAction {
+public:
+  ImageAction();
+
+  enum MirrorPeerType {
```

Comment:
I kind of found it easier when local/remote images maintain (an follow) their respective state transitions -- although for the general add/remove case its pretty much straight forward (i.e., no differentiation), I found it slightly convenient for the shuffle case.